### DISCUSSION TITLE
Some issues while creating a new application in MOOSE

### DISCUSSION THREAD

**Discussion by emdcstanford** [idaholab/moose#18902] [Q&A Getting Started] [answered]
Hi, I recently installed MOOSE on my Mac laptop with the Catalina 10.15.7 operating system. The installation seems to go well, all the tests are passed, and I am able to successfully run the examples, and I also was able to create the "Babbler" example Moose application from the tutorial and run everything. However, I am interested in running a new different application which we called "farm". This application runs without any warnings or errors on a windows computer where it was originally written, but when I do "make -j4" in the new farm application folder on my Mac, I get a number of errors relating to the libmesh folder which strangely seems to be empty in my Moose installation on my Mac. I'm attaching a screenshot showing these errors. It's odd that the function "exp" is not found, I am including cmath in my src files.
<img width="1332" alt="Screen Shot 2021-09-22 at 6 52 59 PM" src="https://user-images.githubusercontent.com/55847507/134444152-d84a78d1-ff4d-412f-97b2-64c3cd4ab784.png">

I should note that this application called "farm" is within the projects folder next to the Moose folder, and I have the .C files in farm/src/kernels/ and the .h files in farm/include/kernels/, and in the installation process I did conda install moose-libmesh. Am I missing something really basic here, or might something have gone wrong in the installation, or are there some compatibility issues from Mac to Windows? In my past experience using MOOSE, there were occasional issues running .i files on my Mac which had been written in a text editor on a Windows. Thank you very much for the help.

**Comment by roystgnr** [comment, marked as answer]
> It's odd that the function "exp" is not found, I am including cmath in my src files.

Including cmath rather than math.h is the right thing to do, but one of the reasons it's the right thing is that it doesn't pollute the global namespace; it puts everything in namespace `std`.  You'll want to either change your `exp` calls to `std::exp` or add a `using std::exp` at the top of your source file.  If a compiler let you get away with a bare exp before then either you were also including math.h elsewhere or you were getting saved by a compiler or library bug somewhere.

I'm kind of surprised to see that `-Werror=reorder` is added to user flags by Moose build.mk; IMHO that's great for Moose to use internally but too nitpicky to force on app developers. ... but git says Moose has been trying to enforce it since 2011, and successfully enforcing it on g++ users since 2013 - maybe you just never hit it before because you were always on Mac and so using a different compiler?  One of these days I'm going to try to talk the Moose people into adopting the libMesh `-Werror` policy (turn on tons of warnings and make them all errors when building our own codes, but don't try to enforce anything on users unless they ask for it), but for now your best bet is to just fix the warning: reorder the initialization lines in your constructor definitions so that they match the order of variables in your class/struct (which is the order in which C++ executes them regardless, IIRC).
